narwhal/consensus/src/tests/randomized_tests.rs

```diff
@@ -0,0 +1,568 @@
+            certificates_per_round
+                .entry(certificate.round())
+                .or_default()
+                .push(certificate.clone());
+
+            certificates.push_back(certificate.clone());
+            next_parents.push(certificate);
+        }
+        parents = next_parents.clone();
+    }
+
+    // sanity check - before return, ensure that we have at least 2f+1 certificates per round and
+    // not certificate exists (except the genesis ones) whose parent is missing.
+    certificates_per_round
+        .iter()
+        .for_each(|(round, round_certs)| {
+            assert!(round_certs.len() >= committee.quorum_threshold() as usize);
+
+            if *round > 0 {
+                let parents = certificates_per_round.get(&(round - 1)).unwrap();
+                round_certs
+                    .iter()
+                    .flat_map(|c| c.header.parents().clone())
+                    .for_each(|digest| {
+                        parents
+                            .iter()
+                            .find(|c| c.digest() == digest)
+                            .unwrap_or_else(|| {
+                                panic!(
+                                    "Certificate with digest {} should be found in parents",
+                                    digest
+                                )
+                            });
+                    });
+            }
+        });
+
+    (certificates, next_parents)
+}
+
+/// Creates various execution plans (`test_iterations` in total) by permuting the order we feed the
+/// DAG certificates to consensus and compare the output to ensure is the same.
+fn generate_and_run_execution_plans(
+    original_certificates: VecDeque<Certificate>,
+    test_iterations: u64,
+    committee: Committee,
+    gc_depth: Round,
+    dag_rounds: Round,
+    run_id: u64,
+    modes: FailureModes,
+) {
+    println!(
+        "Running execution plans for run_id {} for rounds={}, committee={}, gc_depth={}, modes={:?}",
+        run_id,
+        dag_rounds,
+        committee.size(),
+        gc_depth,
+        modes
+    );
+
+    let mut executed_plans = HashSet::new();
+    let mut committed_certificates = Vec::new();
+
+    // Create a single store to be re-used across Bullshark instances to avoid hitting
+    // a "too many files open" issue.
+    let store = make_consensus_store(&test_utils::temp_dir());
+
+    for i in 0..test_iterations {
+        // clear store before using for next test
+        store.clear().unwrap();
+
+        let seed = (i + 1) + run_id;
+
+        let plan = create_execution_plan(original_certificates.clone(), seed);
+
+        let hash = plan.hash();
+        if !executed_plans.insert(hash) {
+            println!("Skipping plan with seed {}, same executed already", seed);
+            continue;
+        }
+
+        // Now create a new Bullshark engine
+        const NUM_SUB_DAGS_PER_SCHEDULE: u64 = 100;
```

Comment:
It would be nice to have just one of these defined per test package and referenced universally since this value doesn't need to be different for each test